app/javascript/src/components/Editor.tsx

```diff
@@ -363,6 +375,148 @@ class Editor extends React.Component<EditorProps, EditorState> {
     }
   }
 
+  private onCopy = () => {
+    const { model } = this
+
+    const selectedNodes = model.getSelectedItems().filter(item => {
+      return item instanceof DefaultNodeModel
+    }) as DefaultNodeModel[]
+    const selectedLinks = this.model.getSelectedItems().filter(item => {
+      return item instanceof DefaultLinkModel
+    }) as DefaultLinkModel[]
+
+    this.copiedLinks = selectedLinks
+    this.copiedNodes = selectedNodes
+    this.pastedNodes = []
+    this.pastedLinks = []
+  }
+
+  private onPaste = () => {
+    const { copiedNodes, model, pastedNodes, pastedLinks, copiedLinks } = this
+
+    model.clearSelection()
+
+    for (let node of copiedNodes) {
+      let copiedNode = this.createCopiedNode(
+        node,
+        node.x + offset,
+        node.y + offset
+      )
+      if (copiedNode) {
+        pastedNodes.push(copiedNode)
+      }
+    }
+    for (let link of copiedLinks) {
+      let copiedLink = this.createCopiedLink(link)
+      if (copiedLink) {
+        pastedLinks.push(copiedLink)
+      }
+    }
+
+    for (let node of pastedNodes) {
+      model.addNode(node)
+    }
+    for (let link of pastedLinks) {
+      model.addLink(link)
+    }
+    this.repaint()
+  }
+
+  private getRelatedPorts = (oldLink: DefaultLinkModel) => {
+    const { copiedNodes, pastedNodes } = this
+
+    let ret = []
+    for (let node of copiedNodes) {
+      // Iterates over the nodes that were copied to clipboard and then over their ports, returning them for the pasted links to use
+      let copiedOutPorts = node.getOutPorts()
```

Comment:
Just for my education (I've never worked on storybook), what are "ports" in this context, and where is `getOutPorts` coming from?

Reply:
<img width="471" alt="Screen Shot 2020-02-27 at 8 40 20 AM" src="https://user-images.githubusercontent.com/16651533/75450439-d08c8000-593c-11ea-8370-133b846460fd.png">

The ports are the connecting dots, with the inPorts being where any lines are coming into the node, and outPorts are the points where the lines leave a node. 

This codebase uses the [react-diagrams](https://github.com/projectstorm/react-diagrams) library, and `getOutPorts()` is a method on that library's `DefaultNodeModel` class, where it returns every port on that node that is an "out" port. In this app, I believe all out ports are always on the right side of the node by convention.

Reply:
Ahhh, ok thanks! I googled "getOutPorts" and couldn't find any JS libs with it, but I see it now in react-diagrams 👍

Reply:
Ahhh, ok thanks! I googled "getOutPorts" and couldn't find any JS libs with it, but I see it now in react-diagrams 👍